clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3575,17 +3575,24 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
 
       Header.addParamDesc(SYCLIntegrationHeader::kind_accessor, Info,
                           CurOffset + offsetOf(FD, FieldTy));
+    } else if (isSyclType(FieldTy, SYCLTypeAttr::stream)) {
+      addParam(FD, FieldTy, SYCLIntegrationHeader::kind_stream);
     } else {
-      if (getMethodByName(ClassTy, FinalizeMethodName))
-        addParam(FD, FieldTy, SYCLIntegrationHeader::kind_stream);
-      else {
-        CXXMethodDecl *InitMethod = getMethodByName(ClassTy, InitMethodName);
-        assert(InitMethod && "type must have __init method");
-        const ParmVarDecl *SamplerArg = InitMethod->getParamDecl(0);
-        assert(SamplerArg && "Init method must have arguments");
-        addParam(SamplerArg->getType(), SYCLIntegrationHeader::kind_sampler,
-                 offsetOf(FD, FieldTy));
+      CXXMethodDecl *InitMethod = getMethodByName(ClassTy, InitMethodName);
+      assert(InitMethod && "type must have __init method");
+      const ParmVarDecl *InitArg = InitMethod->getParamDecl(0);
+      assert(InitArg && "Init method must have arguments");
+      QualType T = InitArg->getType();
+      if (isSyclType(FieldTy, SYCLTypeAttr::sampler)) {
+        addParam(T, SYCLIntegrationHeader::kind_sampler, offsetOf(FD, FieldTy));
+        return true;
       }
+      // Add a default case for classes marked with sycl_special_class
+      // attribute.
+      addParam(T,
```

Comment:
I can do that but wouldn't an assert beat the purpose of a default case? My reasoning for the default case was "new special classes will likely have the same logic because we are trying to unify runtime handling (atleast that is my understanding from older conversations when `sycl_special_class `attribute was implemented)" If we shouldn't make this assumption, we should add the assert

Reply:
Hmm, good point.  On the one hand, the whole idea is we don't change the FE each time; on the other hand, if something does slip through it may be more difficult to find the issue by examining the integration header/run-time failure.

I guess if it happens once, we could add the assertion here at that point.